tests/Operation/BulkWriteTest.php

```diff
@@ -267,7 +267,7 @@ public function testUpdateManyUpdateArgumentMissing(): void
     public function testUpdateManyUpdateArgumentTypeCheck($update): void
     {
         $this->expectException(InvalidArgumentException::class);
-        $this->expectExceptionMessageMatches('/Expected \$operations\[0\]\["updateMany"\]\[1\] to have type "array or object" but found "[\w ]+"/');
+        $this->expectExceptionMessageMatches('/(non-empty pipeline for \$operations\[0\]\["updateMany"\]\[1\])|(\$operations\[0\]\["updateMany"\]\[1\] to have type "array or object" but found ".+")/');
```

Comment:
I notice there's still an "array or object" check for `updateMany` and `updateOne`. Does it make sense to combine that with the `! is_first_key_operator && ! is_pipeline` check that appears later?

```
is_document && ! is_first_key_operator && ! is_pipeline
```

Or might that exclude PackedArray? Aggregation operations require an actual array for their pipelines, so I think it's just update and findAndModify ops that might subtlety allow a PackedArray because they currently accept `array|object`.

Reply:
This indeed allows update pipelines to be passed as `PackedArray` instances. However, giving this some thought, the following options are valid:
- document with update operator as first key
- update pipeline (either as `PackedArray`, array, or `Serializable` that serialises to an array

With that in mind, we can refactor this to throw on this condition:
```
(! is_document || ! is_first_key_operator) && ! is_pipeline
```

I've made the change as the expected type (`array|object`) is less useful than the info that the operation requires an update document or update pipeline.